### PR TITLE
Add an implementation of `mirage-crypto-rng-miou` to initialize the RNG with Miou

### PR DESCRIPTION
**Pull request by dinosaure**
This is my first attempt at implementing `mirage-crypto-rng-miou`. The idea of this initialization is to create a process that will periodically refill the random number generator. I've just tested a program that lasts several hours and I haven't observed any memory leakage: the tasks created to refill the generator are cleaned up properly (every second by default).

The initialization creates a witness to the promise that the user must stop before the end of the application (otherwise Miou raises the `Still_has_children` exception). `Mirage_crypto_rng_miou.kill` just cancels the task and finishes all its sub-tasks.

Finally, I've re-implemented `Fortuna` to be domain-friendly. This is because access may involve changes to the state of the random number generator. We need to ensure that several domains can generate random numbers without conflicts (data-race). A mutex is therefore used to protect exclusive access to the state of the generator. I will verify my implementation with `tsan`.

### CODE REVIEW

**Comment by dinosaure** [comment]
I can confirm via TSan that the current implementation `Pfortuna` does not generates data-races and `Fortuna` does.

**Comment by dinosaure** [comment]
I force-pushed the final version which is rechecked with TSan (I found an issue about Miou itself). The documentation was improved and the `rng` task can be run in parallel or concurrently depending on the number of domains available. 

NOTE: On OCaml 5.1, I was unable to compile tests without the `(modes native)`, so I added them to be able to link correctly C stubs. The error is probably related to the bytecode target and, in some ways, `dune` is not available to link correctly C stubs and bytecode executable.

**Comment by hannesm** [comment]
great to see this happening, but I'm rather uncomfortable with the fork of fortuna.ml -- this has a high maintenance cost. any chance to have a single code base (is there noticable performance if we use mutex everywhere?)

**Comment by dinosaure** [comment]
> this has a high maintenance cost. any chance to have a single code base (is there noticable performance if we use mutex everywhere?)

I don't have strong opinion and we probably can just use `Mutex` (instead of `Miou.Mutex`) to be sure that the `Fortuna` engine is domain-safe for any schedulers. But yes, it requires some benchmarks.

**Comment by hannesm** [comment]
> But yes, it requires some benchmarks.

We have `dune build --rel bench/speed.exe && _build/default/bench/speed.exe fortuna` -- which generates random bytes (but there's no entropy collection happening during that phase, so it may be not the right thing).

**Comment by dinosaure** [comment]
The PR is ready to be reviewed and merged :+1:.

File: bench/dune [hunk 1]
```diff
@@ -2,4 +2,4 @@
  (names speed)
  (modules speed)
  (libraries mirage-crypto mirage-crypto-rng mirage-crypto-rng.unix
-   mirage-crypto-pk mirage-crypto-ec))
+   mirage-crypto-pk mirage-crypto-ec mirage-crypto-rng-miou-unix))
```
Comment (hannesm) [inline bench/dune:5]: no, let's please keep bench/speed.exe working on 4.14, and without any IO library.

you're welcome to contribute a miou-bench.exe.

File: bench/speed.ml [hunk 2]
```diff
@@ -404,6 +404,18 @@ let benchmarks = [
...
+  bm "pfortuna" (fun name ->
+    let open Mirage_crypto_rng_miou_unix.Pfortuna in
+    Miou_unix.run ~domains:2 @@ fun () ->
+    let rng = Mirage_crypto_rng_miou_unix.(initialize (module Pfortuna)) in
+    let g = create () in
+    reseed ~g "abcd" ;
+    throughput name (fun buf ->
+        let buf = Bytes.unsafe_of_string buf in
+        generate_into ~g buf ~off:0 (Bytes.length buf));
+    Mirage_crypto_rng_miou_unix.kill rng) ;
```
Comment (hannesm) [inline bench/speed.ml:417]: see above, this is not the file where miou (or lwt or async or eio) stuff will go. also, above you added an empty file "bench/fortuna.ml"..

**Comment by hannesm** [comment]
looks mostly fine, I had some minor comments. thanks for your contribution.

**Comment by dinosaure** [comment]
It seems that `test_entropy.exe` fails on MacOS (and it's unrelated to this PR). Otherwise, I splitted out the benchmark about `pfortuna` and added a comment about why we re-implemented the module.

**Comment by hannesm** [comment]
> It seems that `test_entropy.exe` fails on MacOS (and it's unrelated to this PR).

Indeed, tracked as #216

**Comment by hannesm** [comment]
Thanks, I rebased on main and force-pushed, after merging #232 there was the only conflict in .cirrus.yml. Also, I added a comment to keep fortuna and pfortuna in sync :)

**Comment by hannesm** [comment]
Once CI passes (see notes in https://github.com/mirage/mirage-crypto/pull/232#issuecomment-2156509775), I'll go ahead and squash-merge this PR.

**Comment by hannesm** [comment]
Hm, I've no clue what to do with bench/miou -- marking it optional doesn't lead to the OCaml-CI not building it -- and marking it with `(package mirage-crypto-rng-miou-unix)` results in a dune error since there's no public_name -- but we don't want a public_name since we don't want it to be installed when installing mirage-crypto-rng-miou-unix.

Any pointers how to get out of this situation? Maybe commenting the dune rule out for the time being?

**Comment by hannesm** [comment]
> Maybe commenting the dune rule out for the time being?

This is what I pushed to this PR. CI looks good. I'll squash & merge if @dinosaure is happy with this PR (esp. with my changes).

**Comment by hannesm** [comment]
> NOTE: On OCaml 5.1, I was unable to compile tests without the `(modes native)`, so I added them to be able to link correctly C stubs. The error is probably related to the bytecode target and, in some ways, `dune` is not available to link correctly C stubs and bytecode executable.

Please note, I had this issue as well, the problem is that if you have mirage-crypto installed, it uses in bytecode mode the installed C stubs instead of the locally compiled ones... And I've a deja-vu https://github.com/ocaml/dune/issues/9979

So, let's remove the `(modes native)`. It is a dune issue, but our tests shouldn't be a scapegoat for that.

**Comment by dinosaure** [comment]
I'm ok with your changes :+1:

**Pull request merged**